NewModuleTest: Don't call both check_jacobian and gradcheck

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49566 NewModuleTest: Don't call both check_jacobian and gradcheck**

Fixes #49422.

check_jacobian and gradcheck do roughly the same thing: they both
compute an analytic jacobian and a numeric jacobian and check that
they are equivalent. Furthermore, NewModuleTest will (by default) call
both check_jacobian and gradcheck, leading to some redundant checks that
waste CI resources.

However, there is one subtle difference: `check_jacobian` can handle the
special case where a Module takes in dense inputs and dense parameters
but returns sparse gradients, but that is not something gradcheck can
handle. This is only used in the tests for nn.Embedding and
nn.EmbeddingBag.

This PR does the following:
- have NewModuleTest call gradcheck instead of check_jacobian by default
- add a new "has_sparse_gradients" flag to NewModuleTest. These are True
for the nn.Embedding and nn.EmbeddingBag sparse gradient tests. If
`has_sparse_gradients` is True, then we call check_jacobian, otherwise,
we call gradcheck.
- Kills the "jacobian_input" flag. This flag was used to tell
NewModuleTest to not attempt to compute the jacobian for the inputs to
the module. This is only desireable if the input to the module isn't
differentiable and was only set in the case of nn.Embedding /
nn.EmbeddingBag that take a LongTensor input. `gradcheck` handles these
automatically by not checking gradients for non-differentiable inputs.

Test Plan:
- Code reading
- run test_nn.py

Differential Revision: [D25622929](https://our.internmc.facebook.com/intern/diff/D25622929)

zou3519 committed:
NewModuleTest: Don't call both check_jacobian and gradcheck

Fixes #49422.

check_jacobian and gradcheck do roughly the same thing: they both
compute an analytic jacobian and a numeric jacobian and check that
they are equivalent. Furthermore, NewModuleTest will (by default) call
both check_jacobian and gradcheck, leading to some redundant checks that
waste CI resources.

However, there is one subtle difference: `check_jacobian` can handle the
special case where a Module takes in dense inputs and dense parameters
but returns sparse gradients, but that is not something gradcheck can
handle. This is only used in the tests for nn.Embedding and
nn.EmbeddingBag.

This PR does the following:
- have NewModuleTest call gradcheck instead of check_jacobian by default
- add a new "has_sparse_gradients" flag to NewModuleTest. These are True
for the nn.Embedding and nn.EmbeddingBag sparse gradient tests. If
`has_sparse_gradients` is True, then we call check_jacobian, otherwise,
we call gradcheck.
- Kills the "jacobian_input" flag. This flag was used to tell
NewModuleTest to not attempt to compute the jacobian for the inputs to
the module. This is only desireable if the input to the module isn't
differentiable and was only set in the case of nn.Embedding /
nn.EmbeddingBag that take a LongTensor input. `gradcheck` handles these
automatically by not checking gradients for non-differentiable inputs.

Test Plan:
- Code reading
- run test_nn.py

[ghstack-poisoned]

zou3519 committed:
Update on "NewModuleTest: Don't call both check_jacobian and gradcheck"


Fixes #49422.

check_jacobian and gradcheck do roughly the same thing: they both
compute an analytic jacobian and a numeric jacobian and check that
they are equivalent. Furthermore, NewModuleTest will (by default) call
both check_jacobian and gradcheck, leading to some redundant checks that
waste CI resources.

However, there is one subtle difference: `check_jacobian` can handle the
special case where a Module takes in dense inputs and dense parameters
but returns sparse gradients, but that is not something gradcheck can
handle. This is only used in the tests for nn.Embedding and
nn.EmbeddingBag.

This PR does the following:
- have NewModuleTest call gradcheck instead of check_jacobian by default
- add a new "has_sparse_gradients" flag to NewModuleTest. These are True
for the nn.Embedding and nn.EmbeddingBag sparse gradient tests. If
`has_sparse_gradients` is True, then we call check_jacobian, otherwise,
we call gradcheck.
- Kills the "jacobian_input" flag. This flag was used to tell
NewModuleTest to not attempt to compute the jacobian for the inputs to
the module. This is only desireable if the input to the module isn't
differentiable and was only set in the case of nn.Embedding /
nn.EmbeddingBag that take a LongTensor input. `gradcheck` handles these
automatically by not checking gradients for non-differentiable inputs.

Test Plan:
- Code reading
- run test_nn.py

Differential Revision: [D25622929](https://our.internmc.facebook.com/intern/diff/D25622929)

[ghstack-poisoned]

mruberry reviewed (commented):
This looks correct to me but I'm not intimately familiar with the two functions at issue, so I'll defer to @albanD to review.

albanD commented on torch/testing/_internal/common_nn.py:
```diff
@@ -5072,7 +5085,8 @@ def _do_test(self, test_case, module, input):
             grad = output.data.clone().normal_()
-            input.grad.data.zero_()
+            if input.grad is not None:
+                input.grad.data.zero_()
```
nit: we should use `torch.no_grad()` or `detach()` instead of `.data`. Not introduced in this PR though so fine if you don't fix it.

zou3519 replied:
Good point,  I will update (or at least try to and see if anything fails)

albanD commented on torch/testing/_internal/common_nn.py:
```diff
@@ -5033,18 +5028,36 @@ def __init__(self, *args, **kwargs):
+        # and nn.EmbeddingBag. Instead, we call `self.check_jacobian`, which
+        # is a slightly different version of gradcheck that can handle this.
+        if self.has_sparse_gradients:
+            assert len(input_tuple) == 1
```
nit `assert num_inputs == 1`

albanD commented on torch/testing/_internal/common_nn.py:
```diff
@@ -5033,18 +5028,36 @@ def __init__(self, *args, **kwargs):
+        # is a slightly different version of gradcheck that can handle this.
+        if self.has_sparse_gradients:
+            assert len(input_tuple) == 1
+            test_input_jacobian = torch.is_floating_point(input_tuple[0])
```
This is a 1:1 correspondence with the flag that used to exit in each definition?

zou3519 replied:
Yes, all of the tests that used to have `test_input_jacobian=True` had non-floating-point inputs.

albanD commented on torch/testing/_internal/common_nn.py:
```diff
@@ -5033,18 +5028,36 @@ def __init__(self, *args, **kwargs):
+        if self.has_sparse_gradients:
+            assert len(input_tuple) == 1
+            test_input_jacobian = torch.is_floating_point(input_tuple[0])
+            test_case.check_jacobian(module, input_tuple[0], test_input_jacobian)
```
nit: the previous version was careful to always pass a tuple as 2nd argument here, but you unpack. Is that the same?

zou3519 replied:
Good catch! Yes, it should be the same. The type annotation for check_jacobian accepts `input` as either a Tensor or a Tuple of tensors:
```
check_jacobian(self, module, input: _TensorOrTensors, jacobian_input=True)
```
In fact, `check_jacobian` actually (around 4 months ago) used to only accept input as Tensor, but we added the ability for it to support passing in `input` as a Tensor and as a tuple of Tensors in https://github.com/pytorch/pytorch/pull/44853

albanD reviewed (approved):
Looks good! Thanks for taking the time to fix this!

zou3519 committed:
Update on "NewModuleTest: Don't call both check_jacobian and gradcheck"


Fixes #49422.

check_jacobian and gradcheck do roughly the same thing: they both
compute an analytic jacobian and a numeric jacobian and check that
they are equivalent. Furthermore, NewModuleTest will (by default) call
both check_jacobian and gradcheck, leading to some redundant checks that
waste CI resources.

However, there is one subtle difference: `check_jacobian` can handle the
special case where a Module takes in dense inputs and dense parameters
but returns sparse gradients, but that is not something gradcheck can
handle. This is only used in the tests for nn.Embedding and
nn.EmbeddingBag.

This PR does the following:
- have NewModuleTest call gradcheck instead of check_jacobian by default
- add a new "has_sparse_gradients" flag to NewModuleTest. These are True
for the nn.Embedding and nn.EmbeddingBag sparse gradient tests. If
`has_sparse_gradients` is True, then we call check_jacobian, otherwise,
we call gradcheck.
- Kills the "jacobian_input" flag. This flag was used to tell
NewModuleTest to not attempt to compute the jacobian for the inputs to
the module. This is only desireable if the input to the module isn't
differentiable and was only set in the case of nn.Embedding /
nn.EmbeddingBag that take a LongTensor input. `gradcheck` handles these
automatically by not checking gradients for non-differentiable inputs.

Test Plan:
- Code reading
- run test_nn.py

Differential Revision: [D25622929](https://our.internmc.facebook.com/intern/diff/D25622929)

[ghstack-poisoned]

zou3519 commented:
There is very interestingly[ a failure on the ROCM build](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.9-py3.6-test1/5790/console):
```
16:46:58   test_ReplicationPad2d_alert_nondeterministic_cuda (__main__.TestNN) ... ok (0.108s)
16:46:58   test_ReplicationPad2d_cuda (__main__.TestNN) ... ok (0.081s)
16:46:58   test_ReplicationPad3d (__main__.TestNN) ... ok (0.466s)
16:46:58   test_ReplicationPad3d_alert_nondeterministic_cuda (__main__.TestNN) ... terminate called after throwing an instance of 'std::out_of_range'
16:46:58   what():  vector::_M_range_check: __n (which is 18446744073709551614) >= this->size() (which is 4)
16:46:59 Traceback (most recent call last):
16:46:59   File "test/run_test.py", line 905, in <module>
16:46:59     main()
16:46:59   File "test/run_test.py", line 888, in main
16:46:59     raise RuntimeError(err_message)
16:46:59 RuntimeError: test_nn failed! Received signal: SIGIOT
16:46:59 
16:46:59 real	61m14.965s
16:46:59 user	243m48.419s
16:46:59 sys	37m44.586s
16:46:59 + cleanup
```

albanD commented:
We should run the tests with `TORCH_SHOW_CPP_STACKTRACES=1`...

Not sure where it could be coming from with just this...
But this PR just disable some tests right? It doesn't run anything that wasn't run before>

zou3519 commented:
> We should run the tests with `TORCH_SHOW_CPP_STACKTRACES=1`...

That sounds useful! I've never tried it before

> Not sure where it could be coming from with just this...
> But this PR just disable some tests right? It doesn't run anything that wasn't run before?

It depends. Some of the tests used to do (check_jacobian, gradcheck, gradgradcheck), but a number of them used to only do (check_jacobian) and this PR switches those over to use (gradcheck). 

But in the case of ReplicationPad3d it doesn't run anything that wasn't run before: previously, (check_jacobian, gradcheck, gradgradcheck) were all run and now only  (gradcheck, gradgradcheck) are run. I will do some digging...

zou3519 commented:
After some investigation, it looks like the error message results from a flakey test: https://ezyang.github.io/pytorch-ci-hud/build1/pytorch-master

Sample runs with similar error messages:
- https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.10-py3.6-test2/294/console
- https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.10-py3.6-test1/110//console

zou3519 committed:
Update on "NewModuleTest: Don't call both check_jacobian and gradcheck"


Fixes #49422.

check_jacobian and gradcheck do roughly the same thing: they both
compute an analytic jacobian and a numeric jacobian and check that
they are equivalent. Furthermore, NewModuleTest will (by default) call
both check_jacobian and gradcheck, leading to some redundant checks that
waste CI resources.

However, there is one subtle difference: `check_jacobian` can handle the
special case where a Module takes in dense inputs and dense parameters
but returns sparse gradients, but that is not something gradcheck can
handle. This is only used in the tests for nn.Embedding and
nn.EmbeddingBag.

This PR does the following:
- have NewModuleTest call gradcheck instead of check_jacobian by default
- add a new "has_sparse_gradients" flag to NewModuleTest. These are True
for the nn.Embedding and nn.EmbeddingBag sparse gradient tests. If
`has_sparse_gradients` is True, then we call check_jacobian, otherwise,
we call gradcheck.
- Kills the "jacobian_input" flag. This flag was used to tell
NewModuleTest to not attempt to compute the jacobian for the inputs to
the module. This is only desireable if the input to the module isn't
differentiable and was only set in the case of nn.Embedding /
nn.EmbeddingBag that take a LongTensor input. `gradcheck` handles these
automatically by not checking gradients for non-differentiable inputs.

Test Plan:
- Code reading
- run test_nn.py

Differential Revision: [D25622929](https://our.internmc.facebook.com/intern/diff/D25622929)

[ghstack-poisoned]
